agent-connection/src/main/kotlin/com/luxoft/blockchainlab/corda/hyperledger/indy/PythonRefAgentConnection.kt

```diff
@@ -0,0 +1,233 @@
+
+    override fun receiveCredential(): CredentialInfo = webSocket.waitForTypedMessage()
+    override fun sendProofRequest(request: ProofRequest) = webSocket.sendTypedMessage(request, counterParty!!)
+
+    override fun receiveProofRequest(): ProofRequest = webSocket.waitForTypedMessage()
+
+    override fun sendProof(proof: ProofInfo) = webSocket.sendTypedMessage(proof, counterParty!!)
+
+    override fun receiveProof(): ProofInfo = webSocket.waitForTypedMessage()
+}
+
+class AgentWebSocketClient(serverUri: URI) : WebSocketClient(serverUri) {
+    private val log = KotlinLogging.logger {}
+
+    override fun onOpen(handshakedata: ServerHandshake?) {
+        log.info { "AgentConnection opened: $handshakedata" }
+    }
+
+    override fun onClose(code: Int, reason: String?, remote: Boolean) {
+        log.info { "AgentConnection closed: $code,$reason,$remote" }
+    }
+
+    val receivedMessages = mutableListOf<String>()
+
+    override fun onMessage(message: String?) {
+        log.info { "Message: $message" }
+        if (message != null)
+            synchronized(receivedMessages) { receivedMessages.add(message) }
+    }
+
+    override fun onError(ex: Exception?) {
+        log.warn(ex) { "AgentConnection error" }
+    }
+
+    fun sendJson(obj: Any) = send(SerializationUtils.anyToJSON(obj))
+
+    fun popMessageOfType(type: String): String? {
+        synchronized(receivedMessages) {
+            val result = receivedMessages.find { SerializationUtils.jSONToAny<Map<String, Any>>(it)["@type"].toString().contentEquals(type) }
+            if (result != null)
+                receivedMessages.remove(result)
+
+            return result
+        }
+    }
+
+    fun waitForMessageOfType(type: String): String {
+        var messageOfType: String? = null
+        while (messageOfType == null) {
+            sleep(500)
+            messageOfType = popMessageOfType(type)
+        }
+        return messageOfType
+    }
+
+    inline fun <reified T : Any> waitForMessageOfType(type: String): T {
+        return SerializationUtils.jSONToAny(waitForMessageOfType(type))
+    }
+
+    fun sendTypedMessage(message: TypedBodyMessage, counterParty: IndyParty) = sendJson(SendMessage(counterParty.did, message))
+    inline fun <reified T : Any> sendTypedMessage(message: T, counterParty: IndyParty) = sendTypedMessage(TypedBodyMessage(message, T::class.java.canonicalName), counterParty)
+    inline fun <reified T : Any> popTypedMessage(): T? {
+        synchronized(receivedMessages) {
+            val message = receivedMessages
+                    .filter { SerializationUtils.jSONToAny<Map<String, Any>>(it)["@type"].toString().contentEquals(MESSAGE_TYPES.MESSAGE_RECEIVED) }
+                    .find { SerializationUtils.jSONToAny<MessageReceived>(it).message.content.clazz == T::class.java.canonicalName }
+
+            if (message != null) {
+                val result = SerializationUtils.jSONToAny<T>(
+                        SerializationUtils.anyToJSON(
```

Comment:
Could we simplify that, make it without back-and-force json mapping?

Reply:
We should use jacksonObjectMapper().convertValue(value) instead

Reply:
Let's create an issue and do in a next PR, please